lldb/source/Plugins/Protocol/MCP/Protocol.cpp

```diff
@@ -0,0 +1,203 @@
+  llvm::json::ObjectMapper O(V, P);
+  if (!O || !O.map("method", N.method))
+    return false;
+  auto *Obj = V.getAsObject();
+  if (!Obj)
+    return false;
+  if (auto *Params = Obj->get("params"))
+    N.params = *Params;
+  return true;
+}
+
+llvm::json::Value toJSON(const ToolCapability &TC) {
+  return llvm::json::Object{{"listChanged", TC.listChanged}};
+}
+
+bool fromJSON(const llvm::json::Value &V, ToolCapability &TC,
+              llvm::json::Path P) {
+  llvm::json::ObjectMapper O(V, P);
+  return O && O.map("listChanged", TC.listChanged);
+}
+
+llvm::json::Value toJSON(const Capabilities &C) {
+  return llvm::json::Object{{"tools", C.tools}};
+}
+
+bool fromJSON(const llvm::json::Value &V, Capabilities &C, llvm::json::Path P) {
+  llvm::json::ObjectMapper O(V, P);
+  return O && O.map("tools", C.tools);
+}
+
+llvm::json::Value toJSON(const TextContent &TC) {
+  return llvm::json::Object{{"type", "text"}, {"text", TC.text}};
+}
+
+bool fromJSON(const llvm::json::Value &V, TextContent &TC, llvm::json::Path P) {
+  llvm::json::ObjectMapper O(V, P);
+  return O && O.map("text", TC.text);
+}
+
+llvm::json::Value toJSON(const TextResult &TR) {
+  return llvm::json::Object{{"content", TR.content}, {"isError", TR.isError}};
+}
+
+bool fromJSON(const llvm::json::Value &V, TextResult &TR, llvm::json::Path P) {
+  llvm::json::ObjectMapper O(V, P);
+  return O && O.map("content", TR.content) && O.map("isError", TR.isError);
+}
+
+llvm::json::Value toJSON(const ToolDefinition &TD) {
+  llvm::json::Object Result{{"name", TD.name}};
+  if (TD.description)
+    Result.insert({"description", TD.description});
+  if (TD.inputSchema)
+    Result.insert({"inputSchema", TD.inputSchema});
+  return Result;
+}
+
+bool fromJSON(const llvm::json::Value &V, ToolDefinition &TD,
+              llvm::json::Path P) {
+
+  llvm::json::ObjectMapper O(V, P);
+  if (!O || !O.map("name", TD.name) ||
+      !O.mapOptional("description", TD.description))
+    return false;
+  return mapRaw(V, "inputSchema", TD.inputSchema, P);
+}
+
+llvm::json::Value toJSON(const Message &M) {
+  return std::visit([](auto &M) { return toJSON(M); }, M);
+}
+
+bool fromJSON(const llvm::json::Value &V, Message &M, llvm::json::Path P) {
+  const auto *O = V.getAsObject();
+  if (!O) {
+    P.report("expected object");
+    return false;
+  }
+
+  if (!O->get("jsonrpc")) {
+    P.report("not a valid JSON RPC message");
+    return false;
+  }
+
+  // A message without an ID is a Notification.
+  if (!O->get("id")) {
+    protocol::Notification N;
+    if (!fromJSON(V, N, P))
+      return false;
+    M = std::move(N);
+    return true;
+  }
+
+  // A message with an error field is an Error.
+  if (O->get("error")) {
+    protocol::Error E;
+    if (!fromJSON(V, E, P))
+      return false;
+    M = std::move(E);
+    return true;
+  }
+
+  // Treat every other message as a request.
+  protocol::Request R;
+  if (!fromJSON(V, R, P))
+    return false;
+  M = std::move(R);
+  return true;
```

Comment:
I think there is one other case for `Response` types.

If the `result` field exists, then its a `Response`.

Also, the `Request` requires the `method` field to exist, otherwise we have a malformed request.